### PR TITLE
Avoid a volatile datasharing reader blocks a writer [12222]

### PR DESCRIPTION
**Pull request by richiware**

### CODE REVIEW

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 1]
```diff
@@ -56,6 +56,18 @@ class PubSubHistory : public testing::TestWithParam<test_params>
...
         mem_policy_ = std::get<1>(GetParam());
+
+        switch (mem_policy_)
+        {
+            case rtps::PREALLOCATED_MEMORY_MODE:
+                will_use_datasharing = true;
+                break;
+            case rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE:
+                will_use_datasharing = true;
+                break;
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:67]:
```suggestion
            case rtps::PREALLOCATED_MEMORY_MODE:
            case rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE:
                will_use_datasharing = enable_datasharing;
                break;
```

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 2]
```diff
@@ -1144,6 +1159,84 @@ TEST_P(PubSubHistory, WriterUnmatchClearsHistory)
...
+    data.index(1u);
+    ASSERT_TRUE(writer.send_sample(data));
+
+    reader.wait_discovery();
+
+    // Second writer sends one sample (reader should discard previous one)
+    data.index(2u);
+    uint32_t expected_value = data.index();
+
+    if (enable_datasharing && will_use_datasharing)
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:1207]:
```suggestion
    if (will_use_datasharing)
```

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 3]
```diff
@@ -1144,6 +1159,84 @@ TEST_P(PubSubHistory, WriterUnmatchClearsHistory)
...
+        {
+            ASSERT_TRUE(writer.send_sample(data));
+        }
+    }
+    else
+    {
+        ASSERT_TRUE(writer.send_sample(data));
+    }
+
+    if (enable_datasharing && will_use_datasharing)
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:1224]:
```suggestion
    if (will_use_datasharing)
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by richiware** [comment]
@MiguelCompany Fixed compilation error

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI!

**Review by MiguelCompany — APPROVED** [review]
LGTM. Failures seem unrelated.

**Comment by richiware** [comment]
Failed tests not related

**Comment by MiguelCompany** [comment]
@mergifyio backport 2.3.x

**Pull request merged**